common/persistence/client/fx.go

```diff
@@ -206,7 +207,10 @@ func managerProvider[T persistence.Closeable](newManagerFn func(Factory) (T, err
 			var nilT T
 			return nilT, err
 		}
-		lc.Append(fx.StopHook(manager.Close))
+		v := reflect.ValueOf(manager)
+		if !v.IsNil() {
```

Comment:
does it work to write it as
```
  if manager != T(nil) ...
```
?
I guess not because T isn't necessarily an interface type? (persistence.Closeable is a type constraint here, not an interface)
I wonder if there's a way to enforce that T is actually an interface type that implements persistence.Closeable?

oh, but this probably works, right?
```
  var nilT T
  if manager != nilT ...
```
(move it out of the error clause)

Reply:
Yeah I tried that but I get `Invalid operation: manager != nilT (the operator != is not defined on T)` so I ended up defining a util method to check if a generic value is nil using reflection.

Reply:
just in case anyone looks at this in the future, one way to do this without reflection is:
```
type comparableCloseable interface {
	comparable
	persistence.Closeable
}

func managerProvider[T comparableCloseable](newManagerFn func(Factory) (T, error)) func(Factory, fx.Lifecycle) (T, error) {
	return func(f Factory, lc fx.Lifecycle) (T, error) {
		manager, err := newManagerFn(f) // passing receiver (Factory) as first argument.
		var nilT T
		if err != nil {
			return nilT, err
		}
		if manager != nilT {
			lc.Append(fx.StopHook(manager.Close))
		}
		return manager, nil
	}
}
```
(I'm still not sure if there's a way to enforce that it's an interface type itself but comparable is close enough)